### PR TITLE
📖 Update optional fields in docs for Machine and Cluster controllers

### LINKED ISSUE #7300: Document `failureDomains` for Cluster controllers in the CAPI Book
/kind bug

The [infra cluster provider contract](https://cluster-api.sigs.k8s.io/developer/providers/cluster-infrastructure.html) lists `failureDomains` as a required contract field. It's not listed in the [cluster controller](https://cluster-api.sigs.k8s.io/developer/architecture/controllers/cluster.html) section however, and it should be added in order to be consistent.

### PR DESCRIPTION
**Pull request by Jont828**
**What this PR does / why we need it**: Update optional fields in docs for Machine and Cluster controllers
**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #7300 #7301

### CODE REVIEW

**Comment by oscr** [comment]
/lgtm 

Content looks accurate based on the code.

Not related to this pr but a general question. How detailed should the documentation be? When it becomes this detailed it's getting very close to just reading the code or the godoc. Comparing:
```
FailureDomainSpec` is defined as:
- `controlPlane` (bool): indicates if failure domain is appropriate for running control plane instances.
- `attributes` (`map[string]string`): arbitrary attributes for users to apply to a failure domain.
```

with https://pkg.go.dev/sigs.k8s.io/cluster-api@v1.2.2/api/v1alpha4#FailureDomainSpec

**Review by killianmuldoon — COMMENTED** [review]
/lgtm

IMO having these definitions be close to the godoc isn't necessarily a bad thing as the godoc isn't a great source of information for the contract definition. 

I think these pages are close to the source of truth for the reality of the provider contracts, so having field information that's close to the level of detail of the godoc isn't IMO an issue.

**Comment by fabriziopandini** [comment]
/lgtm
@sbueringer PTAL

File: docs/book/src/developer/architecture/controllers/cluster.md [hunk 1]
```diff
@@ -44,6 +44,11 @@ The `status` object **may** define several fields that do not affect functionali
 
 * `failureReason` - is a string that explains why a fatal error has occurred, if possible.
 * `failureMessage` - is a string that holds the message contained by the error.
+* `failureDomains` - is a `failureDomains` type indicating the failure domains that machines should be placed in. `failureDomains`
```
Comment (sbueringer) [inline docs/book/src/developer/architecture/controllers/cluster.md:47]:
```suggestion
* `failureDomains` - is a `FailureDomains` type indicating the failure domains that machines should be placed in. `FailureDomains`
```
nit: please also adjust in cluster-infrastructure.md

File: docs/book/src/developer/architecture/controllers/cluster.md [hunk 2]
```diff
@@ -44,6 +44,11 @@ The `status` object **may** define several fields that do not affect functionali
...
+is a map, defined as `map[string]FailureDomainSpec`. A unique key must be used for each `FailureDomainSpec`.
+`FailureDomainSpec` is defined as:
+- `controlPlane` (bool): indicates if failure domain is appropriate for running control plane instances.
```
Comment (sbueringer) [inline docs/book/src/developer/architecture/controllers/cluster.md:50]: I think the indentation is off. This should be one level deeper

File: docs/book/src/developer/architecture/controllers/machine.md [hunk 3]
```diff
@@ -81,6 +81,12 @@ The `spec` object **must** at least one field defined:
 
 * `providerID` - a cloud provider ID identifying the machine.
 
+#### Optional `spec` fields
+
+The `spec` object **may** define several fields that do not affect functionality if missing:
+
+* `failureDomain` - is a string indentifying the failure domain the instance is running in.
```
Comment (sbueringer) [inline docs/book/src/developer/architecture/controllers/machine.md:88]:
```suggestion
* `failureDomain` - is a string identifying the failure domain the instance is running in.
```
nit

File: docs/book/src/developer/architecture/controllers/machine.md [hunk 4]
```diff
@@ -93,6 +99,11 @@ The `status` object **may** define several fields that do not affect functionali
 
 * `failureReason` - is a string that explains why a fatal error has occurred, if possible.
 * `failureMessage` - is a string that holds the message contained by the error.
+* `addresses` - is a list of `MachineAddress` which represents host names, external IP addresses, internal IP addresses,
+external DNS names, and/or internal DNS names for the provider's machine instance. `MachineAddress` is
+defined as:
+    - `type` (string): one of `Hostname`, `ExternalIP`, `InternalIP`, `ExternalDNS`, `InternalDNS`
+    - `address` (string)
```
Comment (sbueringer) [inline docs/book/src/developer/architecture/controllers/machine.md:106]: Looks good. Just saw that the indentation is off in https://cluster-api.sigs.k8s.io/developer/providers/machine-infrastructure.html

Could you please fix the indentation there? Thank you very much!

File: docs/book/src/developer/providers/machine-infrastructure.md [hunk 5]
```diff
@@ -40,8 +40,8 @@ A machine infrastructure provider must define an API type for "infrastructure ma
         3. `addresses` (`MachineAddress`): a list of the host names, external IP addresses, internal IP addresses,
```
Comment (sbueringer) [inline docs/book/src/developer/providers/machine-infrastructure.md:40]:
```suggestion
        3. `addresses` (`MachineAddresses`): a list of the host names, external IP addresses, internal IP addresses,
```

File: docs/book/src/developer/architecture/controllers/machine.md [hunk 6]
```diff
@@ -93,6 +99,11 @@ The `status` object **may** define several fields that do not affect functionali
 
 * `failureReason` - is a string that explains why a fatal error has occurred, if possible.
 * `failureMessage` - is a string that holds the message contained by the error.
+* `addresses` - is a list of `MachineAddress` which represents host names, external IP addresses, internal IP addresses,
```
Comment (sbueringer) [inline docs/book/src/developer/architecture/controllers/machine.md:102]:
```suggestion
* `addresses` - is a `MachineAddresses` (a list of `MachineAddress`) which represents host names, external IP addresses, internal IP addresses,
```

**Review by sbueringer — COMMENTED** [review]
Sorry, last nits from my side. Just noticed that the current definition of addresses are not entirely correct

**Comment by oscr** [comment]
@Jont828 Could you please also squash your commits? If you don't know how it's described in [CONTRIBUTING](https://github.com/kubernetes-sigs/cluster-api/blob/main/CONTRIBUTING.md#contributing-a-patch)

**Comment by Jont828** [comment]
@oscr Done, let me know if there's anything else I need to do to get this merged!

**Comment by sbueringer** [comment]
Thank you very much!

/lgtm
/approve

**Pull request merged**